src/lib/vrf_lib/standalone.ml

```diff
@@ -152,28 +154,77 @@ end = struct
 
   module Evaluation = struct
     module Discrete_log_equality = struct
-      type 'scalar t_ = {c: 'scalar; s: 'scalar} [@@deriving sexp, bin_io]
+      module Stable = struct
+        module V1 = struct
+          type 'scalar t_ = {c: 'scalar; s: 'scalar} [@@deriving sexp, bin_io]
+
+          module T = struct
+            let version = 1
 
-      type t = Scalar.t t_ [@@deriving sexp, bin_io]
+            type t = Scalar.t t_ [@@deriving sexp, bin_io]
```

Comment:
This is only safe if `Scalar.t` is versioned, right?

Reply:
`Scalar` is a functor argument, so yes, the serialization depends on that argument. So yes, the argument should be versioned (or never, ever change).

I'd say the changes in this PR are independent of how the functor is applied, though.

Also, I could not find any uses of the functor. I did find uses of the other functor in this library, `Integrated.Make`. Was this one test code, maybe?

Reply:
if we version the `Scalar` argument to the functor, then we also tie the implementation to that version of the Scalar. So `create` (for example) would return a type with `Scalar.V1.t` instead of `Scalar.Latest` -- moreover when we want to V2 this scalar, we'd need to somehow also pass V1 into the functor and still we'd want to use latest in the implementation.

I wonder if we should require:

```ocaml
module Scalar : Versioned.S
```

or something instead of `type t [@@derving bin_io, ...]`

Reply:
If I follow, you mean, force this functor to be applied to a particular `Scalar` version by specifying a version in the functor signature?

I'm not sure that approach works, in general, because you could (in theory) have two `Scalar` implementations that have a particular versioned module, say `V1`, that's mentioned in the signature.

If you want to force a particular version of a particular `Scalar` implementation, would it make more sense to remove it as a functor argument, and just use it, in-line?

Reply:
I'm thinking this would be `Versioned.S1`:

```ocaml
sig
  module Stable : sig
    module V1 : sig
      type t = (* ... *) [@@deriving bin_io]
    end
  end
  type t = (* ... *)
end
```

and `Versioned.S2` could have versions 1 and 2
and `Versioned.S3` could have 1,2,3 etc

Reply:
and yeah we'd have to update all applications of the functor in lockstep if we change only one of them, but I think this is okay.

Reply:
Maybe it's not a real problem, but doesn't this leave you vulnerable to two `Stable` implementations with the same versioned modules, with different types/serializations?

Reply:
How so?

Reply:
Suppose a functor is applied to a versioned type, and the functor signature mentions the versions, as we've done in this PR.

Suppose also that the functor is really using the argument as a parameter that can vary, so it's applied to two different versioned types.

Issue 1: the choice of version is made inside the functor, so that if it chooses, say V2, we'd have to choose V2 for both versioned types. But those types are distinct, and V2 of one has no particular relation to V2 of the other. V2 might not exist for both.

Issue 2: Applying the functor to the two different versioned types yields two different versioned types, whose serializations are distinct.  The versioned type inside the functor does not identify a unique type. This may be OK, as long as we keep this fact in mind, and don't confuse the two serializations.

It looks like, most of the time, functors are applied to a single argument. That's true in this PR. We can remove these issues by simply inlining the type we were going use as an argument. In that case, the versioned type inside the functor is unique with a particular serialization.

Reply:
I'm not sure I follow this, could you provide a code snippet to explain?

Reply:
basically if you change this line to `Scalar.V1.t t_` it should compile I think

Reply:
Part of what I was thinking is an issue, is not really an issue.

But still:

Functor that takes a versioned module:
```
module Make (A : sig ... V1 ... V2 ... end) = ...  (* Versioned.S2, if you will *)
```
The functor arguments:
```
module A1 = ... V1 ... V2 ...
module A2 = ... V1 ...

module Foo = Make (A1)
module Bar = Make (A2)
```
which would fail to compile, because `A2` doesn't haven't a `V2`.

Distinct implementations of versioned modules can diverge in incompatible ways, while the functor signature wants them to be the same.

Also, `Foo` and `Bar` have distinct serializations of any type that uses `A`, even though it's the same-named type. Probably OK, as long as you know that.

Reply:
We'd have to increment them in lockstep whenever only one of them changes I guess